contracts/src/LightClient.sol

```diff
@@ -0,0 +1,157 @@
+        if (genesis.viewNum != 0 || genesis.blockHeight != 0) {
+            revert InvalidArgs();
+        }
+
+        genesisState = genesis;
+        finalizedState = genesis;
+        currentEpoch = 0;
+        stakeTable = IStakeTable(stakeTableAddr);
+        // TODO: (alex) initialized stake table or at least store its contract address ref here
+    }
+
+    // === State Modifying APIs ===
+    //
+    /// @notice Update the latest finalized light client state. It is expected to be updated
+    /// periodically, especially an update for the last block for every epoch has to be submitted
+    /// before any newer state can be accepted since the stake table commitments of that block
+    /// become the snapshots used for vote verifications later on.
+    function newFinalizedState(
+        LightClientState calldata newState,
+        IPlonkVerifier.PlonkProof calldata proof
+    ) external {
+        if (
+            newState.viewNum <= finalizedState.viewNum
+                || newState.blockHeight <= finalizedState.blockHeight
+        ) {
+            revert OutdatedState();
+        }
+        uint64 epochEndingBlockHeight = (currentEpoch + 1) * BLOCKS_PER_EPOCH - 1;
+        if (
+            finalizedState.blockHeight != epochEndingBlockHeight
+                && newState.blockHeight > epochEndingBlockHeight
+        ) {
+            revert MissingLastBlockForCurrentEpoch(epochEndingBlockHeight);
+        }
+        // format validity check
+        BN254.validateScalarField(newState.blockComm);
+        BN254.validateScalarField(newState.feeLedgerComm);
+        BN254.validateScalarField(newState.stakeTableBlsKeyComm);
+        BN254.validateScalarField(newState.stakeTableSchnorrKeyComm);
+        BN254.validateScalarField(newState.stakeTableAmountComm);
+        // sanity check on the threshold
+        if (newState.threshold <= stakeTable.totalVotingStake() * 2 / 3) {
+            revert InvalidArgs();
+        }
+
+        // check plonk proof
+        // TODO: (alex) replace the vk with the correct one
+        IPlonkVerifier.VerifyingKey memory vk = VkTest.getVk();
+        uint256[] memory publicInput = preparePublicInput(newState);
+        PlonkVerifier.verify(vk, publicInput, proof, bytes(""));
+
+        // upon successful verification, update state.
+        // If the newState is in a new epoch, only then should we increment the `currentEpoch`, and
+        // update the stake table. The `finalizedState` (before update) should have the
+        // `epochEndingBlockHeight`
+        if (finalizedState.blockHeight == epochEndingBlockHeight) {
+            bytes32 newStakeTableComm = keccak256(
+                abi.encodePacked(
+                    finalizedState.stakeTableBlsKeyComm,
+                    finalizedState.stakeTableSchnorrKeyComm,
+                    finalizedState.stakeTableAmountComm
+                )
+            );
+
+            stakeTable.advanceEpoch(newStakeTableComm);
```

Comment:
Why do we need this? Aren't epochs defined from HotShot honest nodes reaching consensus on the ethereum block that is read when the last block of some Hotshot epoch is generated? Otherwise depending on when `newFinalizedState` is called,  it seems that we may have different definitions of the stake table.

Reply:
Ok I understand, this is just to store the new stake table commitment.